lib/src/recovery_protocol/angular_analyzer_protocol.dart

```diff
@@ -0,0 +1,602 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of angular_ast.src.recovery_protocol.recovery_protocol;
+
+class NgAnalyzerRecoveryProtocol extends RecoveryProtocol {
+  @override
+  RecoverySolution hasError(
+      NgSimpleToken current, NgTokenReversibleReader reader) {
+    return new RecoverySolution.skip();
+  }
+
+  @override
+  RecoverySolution isEndOfFile(
+      NgSimpleToken current, NgTokenReversibleReader reader) {
+    return new RecoverySolution.skip();
+  }
+
+  @override
+  RecoverySolution scanAfterComment(
+      NgSimpleToken current, NgTokenReversibleReader reader) {
+    NgScannerState returnState;
+    NgToken returnToken;
+
+    if (current.type == NgSimpleTokenType.EOF) {
+      reader.putBack(current);
+      returnState = NgScannerState.scanStart;
+      returnToken = new NgToken.generateErrorSynthetic(
+          current.offset, NgTokenType.commentEnd);
+      return new RecoverySolution(returnState, returnToken);
+    }
+    return new RecoverySolution(returnState, returnToken);
+  }
+
+  @override
+  RecoverySolution scanAfterElementDecorator(
+      NgSimpleToken current, NgTokenReversibleReader reader) {
+    NgScannerState returnState;
+    NgToken returnToken;
+    NgSimpleTokenType type = current.type;
+    int offset = current.offset;
+
+    if (type == NgSimpleTokenType.openBracket ||
+        type == NgSimpleTokenType.openParen ||
+        type == NgSimpleTokenType.openBanana ||
+        type == NgSimpleTokenType.hash ||
+        type == NgSimpleTokenType.star ||
+        type == NgSimpleTokenType.closeBracket ||
+        type == NgSimpleTokenType.closeParen ||
+        type == NgSimpleTokenType.closeBanana) {
+      reader.putBack(current);
+      returnState = NgScannerState.scanElementDecorator;
+      returnToken = new NgToken.generateErrorSynthetic(
+          offset, NgTokenType.beforeElementDecorator,
+          lexeme: " ");
+    } else if (type == NgSimpleTokenType.EOF ||
+        type == NgSimpleTokenType.commentBegin ||
+        type == NgSimpleTokenType.openTagStart ||
+        type == NgSimpleTokenType.closeTagStart) {
+      reader.putBack(current);
+      returnState = NgScannerState.scanStart;
+      returnToken = new NgToken.generateErrorSynthetic(
+          offset, NgTokenType.openElementEnd);
+    } else if (type == NgSimpleTokenType.doubleQuote ||
+        type == NgSimpleTokenType.singleQuote) {
+      int _offset = (current is NgSimpleQuoteToken)
+          ? current.quoteOffset
+          : current.offset;
+      reader.putBack(current);
+      returnState = NgScannerState.scanElementDecoratorValue;
+      returnToken = new NgToken.generateErrorSynthetic(
+          _offset, NgTokenType.beforeElementDecoratorValue);
+    }
+
+    return new RecoverySolution(returnState, returnToken);
```

Comment:
Interesting. I didn't even think of it this way (largely due to lack of knowledge of how RegEx works internally). I think it would be cool to see if we can find a more systematic way to implement this type of recovery.